### PR TITLE
Add support for custom coordinate names

### PR DESCRIPTION
**Pull request by huard**
As proposed in #74, this PR uses a dictionary of variable names to indicate which variables hold latitude and longitude coordinates and bounds. If this dictionary is not given, the code uses very basic heuristics based on the CF-Convention to find the coordinates and their bounds. The bounds are then converted into the format required by ESMF. 

Note that the conversion from vertices (nx, ny, 4) is probably wrong and needs additional work and testing.

### CODE REVIEW

**Comment by jthielen** [comment]
Glad to see this moving forward before a full `cf-xarray` package is ready to be integrated!

I have one early suggestion: instead of having the basic CF-based check being against the `long_name` attribute (which is not standardized in the current conventions), have it check for the required `units` attribute
- for latitude, this is one of degrees_north, degree_north, degree_N, degrees_N, degreeN, or degreesN (see http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#latitude-coordinate).
- for longitude, this is one of degrees_east, degree_east, degree_E, degrees_E, degreeE, or degreesE (see http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#longitude-coordinate).

**Comment by huard** [comment]
Done. Just wondering whether this is a one-to-one relationship. Lats and lons have to have those units, but is any variable with those units guaranteed to be --the-- grid coordinate ?

**Comment by huard** [comment]
Another question: there is a potential for confusion between two types of "bounds": 
- the grid corners required by xESMF (n+1, m+1), 
- the grid vertices defined in the CF convention (n, 2) or (n, m, 4)

Suggestions about the terminology to use to make this distinction clear ? The code currently uses *bounds* and *boundaries* interchangeably.

**Comment by jthielen** [comment]
> Just wondering whether this is a one-to-one relationship. Lats and lons have to have those units, but is any variable with those units guaranteed to be --the-- grid coordinate ?

From my reading of the conventions document, while it is not explicitly stated to be one-to-one/uniquely identifying, it is implied since this is how the conventions instruct applications to interpret those coordinate types (and such an instruction wouldn't be very useful if it wasn't uniquely identifying). For example:

> Coordinates of latitude with respect to a rotated pole should be given units of degrees, not degrees_north or equivalents, because applications which use the units to identify axes would have no means of distinguishing such an axis from real latitude, and might draw incorrect coastlines, for instance.

File: xesmf/util.py [hunk 1]
```diff
@@ -96,3 +150,197 @@ def grid_global(d_lon, d_lat):
                       'might not cover the globe uniformally'.format(d_lat))
 
     return grid_2d(-180, 180, d_lon, -90, 90, d_lat)
+
+
+def is_longitude(var):
+    """Return True if variable is a longitude.
+
+    A variable is considered a longitude if its `units` are 'degrees_east' or a synonym.
```
Comment (stefraynaud) [inline xesmf/util.py]: Yes. That's part of the Cf conventions.
Other criteria from [cf paragraph](http://cfconventions.org/cf-conventions/cf-conventions.html#longitude-coordinate): standard_name and axis attributes, despite the checking with axis attribute may lead to errors with 2D coordinates.

File: xesmf/util.py [hunk 2]
```diff
@@ -96,3 +150,197 @@ def grid_global(d_lon, d_lat):
...
+
+    Return
+    ------
+    bool
+      True if variable is a latitude.
+    """
+    return has_units_of_latitude(var)
+
+
+def has_name_of_longitude(var):
```
Comment (stefraynaud) [inline xesmf/util.py]: name -> standard_name?

File: xesmf/util.py [hunk 3]
```diff
@@ -96,3 +150,197 @@ def grid_global(d_lon, d_lat):
...
+    key = "standard_name"
+    return var.attrs.get(key) == "longitude"
```
Comment (stefraynaud) [inline xesmf/util.py]: Be careful, models with staggered grids use standard_name like "longitude_at_u_location".

File: xesmf/util.py [hunk 4]
```diff
@@ -96,3 +150,197 @@ def grid_global(d_lon, d_lat):
...
+    var_names: dict
+      Dictionary with keys `lon` and `lat` identifying the name of the variables storing longitude and latitude
+      coordinates respectively.
+
+    Return
+    ------
+    np.ndarray, np.ndarray
+      Longitude and latitude arrays.
+    """
+    if var_names is None:
```
Comment (stefraynaud) [inline xesmf/util.py]: Should it support the case of an incomplete dictionary, for instance with missing lat or missing names for bounds.
If the dict is provided without name for bounds only, lon and lat are found explicitly and bounds are found from CF checking.

File: xesmf/frontend.py [hunk 5]
```diff
@@ -11,13 +11,20 @@
...
     dask_array_type = (da.Array,)  # for isinstance checks
 except ImportError:
     dask_array_type = ()
 
+default_var_names = {"lat": "lat",
+                     "lon": "lon",
+                     "lat_b": "lat_b",
+                     "lon_b": "lon_b"}
+
+
```
Comment (stefraynaud) [inline xesmf/frontend.py:27]: I was wondering if it wouldn't be appropriate for these defaults to be more easily accessible to the user, e.g. via an `xesmf.config.set` as discussed in #74?

File: xesmf/util.py [hunk 6]
```diff
@@ -28,6 +28,56 @@ def _grid_1d(start_b, end_b, step):
...
+    Returns
+    -------
+    array (n, m, 4)
+      Counter
+    """
+    return np.dstack((bounds[:-1, :-1], bounds[1:, :-1], bounds[1:, 1:], bounds[:-1, 1:]))
+
+
+def cf_to_esm_bnds_1d(vertices):
+    """Convert 2D CF-compliant bounds to 1D ESM boundaries."""
```
Comment (stefraynaud) [inline xesmf/util.py]: This is part of the public API, so you should probably inform the user of the inputs and outputs.

File: xesmf/util.py [hunk 7]
```diff
@@ -28,6 +28,56 @@ def _grid_1d(start_b, end_b, step):
...
+    v = vertices
+    return np.concatenate((v[:, 0], v[-1:, 1]))
+
+
+# TODO: This is probably not true in general, and depends on the convention used, starting point, clockwise, etc.
+def cf_to_esm_bnds_2d(vertices):
+    """Convert 3D CF-compliant bounds to 2D ESM boundaries."""
```
Comment (stefraynaud) [inline xesmf/util.py]: Here too.

File: xesmf/util.py [hunk 8]
```diff
@@ -96,3 +150,197 @@ def grid_global(d_lon, d_lat):
...
+    """
+    return has_units_of_latitude(var)
+
+
+def has_name_of_longitude(var):
+    key = "standard_name"
+    return var.attrs.get(key) == "longitude"
+
+
+def has_name_of_latitude(var):
```
Comment (stefraynaud) [inline xesmf/util.py]: A minimal doc? Well, the name is explicit. Perhaps just to say that the input is a datarray.

File: xesmf/util.py [hunk 9]
```diff
@@ -28,6 +28,56 @@ def _grid_1d(start_b, end_b, step):
...
+    array (n, m, 4)
+      Counter
+    """
+    return np.dstack((bounds[:-1, :-1], bounds[1:, :-1], bounds[1:, 1:], bounds[:-1, 1:]))
+
+
+def cf_to_esm_bnds_1d(vertices):
+    """Convert 2D CF-compliant bounds to 1D ESM boundaries."""
+    v = vertices
+    return np.concatenate((v[:, 0], v[-1:, 1]))
```
Comment (stefraynaud) [inline xesmf/util.py:77]: Perhaps add in the docstring that the cell bounds are contiguous.

File: xesmf/util.py [hunk 10]
```diff
@@ -96,3 +150,197 @@ def grid_global(d_lon, d_lat):
...
+                         "Identify the longitude bounds by setting the `bounds` attribute of the longitude variable "
+                         "to an existing variable name.")
+    if key not in ds:
+        raise ValueError(f"The variable {key} identified as the bounds for {coord.name} is not in this dataset.")
+
+    return ds[key]
+
+
+def cf_lon_lat_bnds(ds):
+    """Return longitude and latitude boundaries.
```
Comment (stefraynaud) [inline xesmf/util.py]: Maybe add a reminder about the array shapes.

File: xesmf/util.py [hunk 11]
```diff
@@ -96,3 +150,197 @@ def grid_global(d_lon, d_lat):
...
+    ----------
+    ds: xr.Dataset
+      Dataset storing coordinate information.
+    var_names: dict
+      Dictionary with keys `lon_b` and `lat_b` identifying the name of the variables storing longitude and latitude
+      coordinate boundaries respectively.
+
+    Return
+    ------
+    np.ndarray, np.ndarray
```
Comment (stefraynaud) [inline xesmf/util.py:431]: With the shapes?

**Review by stefraynaud — APPROVED** [review]
This PR will help users to use xesmf with their own data formats in most cases.
Minor changes are proposed, at least to clarify a few things.

File: xesmf/util.py (hunk 1 above)
Comment (huard) [reply xesmf/util.py]: Now filters on 
```python
has_units_of_longitude(var) or has_standard_name_of_longitude(var)
```

File: xesmf/util.py (hunk 2 above)
Comment (huard) [reply xesmf/util.py]: Done

File: xesmf/util.py (hunk 3 above)
Comment (huard) [reply xesmf/util.py]: Should then this be 
```
var.attrs.get(key).startswith("longitude")
```
?

File: xesmf/util.py (hunk 4 above)
Comment (huard) [reply xesmf/util.py]: Something like 
```
if var_names is None or "lon" not in var_names or "lat" not in var_names:
```
?

File: xesmf/frontend.py (hunk 5 above)
Comment (huard) [reply xesmf/frontend.py:27]: My sense is that these values are file-specific, so setting default values for a *session* seems weird. Can this be done in another PR ?

File: xesmf/util.py (hunk 6 above)
Comment (huard) [reply xesmf/util.py]: Done. Should the naming be more explicit ? vertices, bounds, boundaries are used a bit loosely throughout the current code.

File: xesmf/util.py (hunk 7 above)
Comment (huard) [reply xesmf/util.py]: I'm pretty sure this implementation is not robust. It assumes coordinates are given starting from the upper left corner and going anti-clockwise.

File: xesmf/util.py (hunk 8 above)
Comment (huard) [reply xesmf/util.py]: Done

File: xesmf/util.py (hunk 9 above)
Comment (huard) [reply xesmf/util.py:77]: Done, please check that this is what you meant.

File: xesmf/util.py (hunk 11 above)
Comment (huard) [reply xesmf/util.py:431]: Done

File: xesmf/util.py (hunk 10 above)
Comment (huard) [reply xesmf/util.py]: Done

**Comment by huard** [comment]
Thanks @stefraynaud for the review. The one thing that still bothers me with the PR is the naming convention. There are various terms used that I think we should use more consistently:
- bounds
- boundaries
- vertices
- corners

I don't have any actual suggestion to make to clear this up.

**Comment by stefraynaud** [comment]
As for the terms, maybe it can be clarified in two generic steps:
1. Make sure that the docstrings mentions the shape of arrays.
2. Clarify the user guide about the possible terms.

File: xesmf/util.py (hunk 3 above)
Comment (stefraynaud) [reply xesmf/util.py]: Without regexp, I would suggest an equality comparison to "longitude" and, in addition, that the string starts with "longitude_at_".

File: xesmf/frontend.py (hunk 5 above)
Comment (stefraynaud) [reply xesmf/frontend.py:27]: Yes, that what I was finally thinking.

File: xesmf/util.py (hunk 3 above)
Comment (huard) [reply xesmf/util.py]: Done.

**Comment by huard** [comment]
The code to convert nxmx4 bounds to 2D grid corners should make respect this convention. I don't think it does at the moment: https://docs.google.com/document/d/1ueYsYrPob8Gd3HxQVHC0vKfpj_Jjh0iO4Pe70_QLpDk/edit

**Comment by rabernat** [comment]
If this PR is ready to merge, perhaps it can be moved to https://github.com/pangeo-data/xESMF? Unfortunately this has to be done manually.

**Comment by jthielen** [comment]
Also, given that [cf-xarray](https://github.com/xarray-contrib/cf-xarray) has been released, would it be worth just switching over to depending on that at this point?

**Comment by rabernat** [comment]
We discussed that at the last dev meeting (which you're welcome to join any time; see #98). @huard concluded it might not be worth it in this specific case.

**Comment by jthielen** [comment]
> We discussed that at the last dev meeting (which you're welcome to join any time; see #98). @huard concluded it might not be worth it in this specific case.

Sounds good! That's definitely reasonable for it to not be worth it in this case, and I'm glad to hear it was discussed. I'll be using xESMF and cf-xarray in combination for an upcoming project or two, so if there are any sticking points, I'll be sure to put in a PR to help.

**Comment by dcherian** [comment]
> it might not be worth it in this specific case.

So this PR was a main reason I wrote it up, so that logic to figure lat/lon didn't have to be written 5-10x over. 

The logic is pretty self contained so cf-xarray's utility functions *should* be easily copy/pastable.

**Comment by rabernat** [comment]
FWIW, I also favor outsourcing this to cf-xarray. However, I am not the author of the PR and so must defer to his judgement. Perhaps we could merge this as is, with good test coverage, and then refactor to use cf-xarray in a future PR without touching the tests?

**Comment by stefraynaud** [comment]
Historically, this PR came after PR #73.
This latter was proposed to make the job of cf-xarray, and it has been chosen to keep only #94 because it simpler implementation.
Now that cf-xarray is here, a new PR must be proposed to merge to two capabilities:

* Let the user specify the lon/lat names.
* Find the lon/lat names according to CF conventions.

This means: if the names are not specified, then search for them according to the default names, and if not found, use cf-xarray,

**Comment by huard** [comment]
One concern was weighing the maintenance costs of adding a dependency versus maintaining code. I thought the code was so basic, it was not worth an additional dependency. As xesmf and cf-xarray both evolve, I suspect the tradeoff will lean in favor of relying on cf-xarray. 

I think this PR still has an issue with the conversion from 2D vertices to grid corners however, and I'm short on time to investigate this. Is this in-scope for cf-xarray ?

**Comment by dcherian** [comment]
> One concern was weighing the maintenance costs of adding a dependency versus maintaining code. I thought the code was so basic, it was not worth an additional dependency. As xesmf and cf-xarray both evolve, I suspect the tradeoff will lean in favor of relying on cf-xarray.

another approach is to make the code in cf-xarray easily copy-pastable which it may be already.

> I think this PR still has an issue with the conversion from 2D vertices to grid corners however, and I'm short on time to investigate this. Is this in-scope for cf-xarray ?

Sure. I added a function to guess 1D bounds (edges) already: https://cf-xarray.readthedocs.io/en/latest/generated/xarray.Dataset.cf.add_bounds.html#xarray.Dataset.cf.add_bounds. We can add a 2D version too.

**Comment by huard** [comment]
Great ! If xesmf can rely on cf-xarray for the variable identification and corner grid calculations, I think it would make sense to ditch this PR in favor of a new one using cf-xarray as a dependency.

**Comment by huard** [comment]
@sbiner Voir https://xclim--576.org.readthedocs.build/en/576/notebooks/frequency_analysis.html Est-ce compréhensible ?

**Pull request closed without merging**